--- a/webaudio-tinysynth.js
+++ b/webaudio-tinysynth.js
@@ -701,6 +701,8 @@
               if(this.playIndex>=this.song.ev.length){
                 if(this.loop){
                   e=this.song.ev[this.playIndex=0];
+                  if(this.loopEnd)
+                    this.playTime+=(Math.max(this.loopEnd,this.playTick)-this.playTick+e.t)*this.tick2Time;
                   this.playTick=e.t;
                 }
                 else{
@@ -898,7 +900,7 @@
             case 0x2f:
               return 1;
             case 0x51:
-              var val = Math.floor(60000000 / Get3(s, i + 3));
+              var val = 60000000 / Get3(s, i + 3);
               song.ev.push({t:tick, m:[0xff51, val]});
               break;
             }
